client/src/main/java/org/eclipse/hono/client/impl/GatewayOrDeviceSpecificCommandConsumer.java

```diff
@@ -0,0 +1,344 @@
+        }
+        commandConsumerHandlers.put(handler.getDeviceId(), handler);
+        return Future.succeededFuture(handler.getDeviceId().equals(gatewayOrDeviceId) ? this
+                : new GatewayManagedDeviceCommandConsumer(handler.getDeviceId()));
+    }
+
+    private boolean isValidCommandConsumerHandler(final CommandConsumerHandler handler) {
+        return gatewayOrDeviceId.equals(handler.getGatewayId())
+                || (handler.getDeviceId().equals(gatewayOrDeviceId) && handler.getGatewayId() == null);
+    }
+
+    /**
+     * Checks whether a handler exists for the given device id.
+     * 
+     * @param deviceId The device id.
+     * @return {@code true} if a handler exists.
+     */
+    public boolean containsCommandConsumerHandler(final String deviceId) {
+        return commandConsumerHandlers.containsKey(deviceId);
+    }
+
+    /**
+     * Gets a handler for either the given device id or for the device id of <em>this</em>
+     * command consumer.
+     * 
+     * @param gatewayManagedDeviceId The id of a device connected to a gateway. May be {@code null}.
+     *                               If a handler for that device exists, it will be returned here.
+     *                               Otherwise the handler for <em>this</em> command consumers
+     *                               device id will be returned (if set).
+     * @return The handler or {@code null}.
+     */
+    public CommandConsumerHandler getCommandConsumerHandlerOrDefault(final String gatewayManagedDeviceId) {
+        if (gatewayManagedDeviceId != null && !gatewayManagedDeviceId.equals(gatewayOrDeviceId)) {
+            final CommandConsumerHandler commandConsumerHandler = commandConsumerHandlers.get(gatewayManagedDeviceId);
+            if (commandConsumerHandler != null) {
+                LOG.trace("using device specific command handler [gateway-id: {}, device-id: {}]", gatewayOrDeviceId,
+                        gatewayManagedDeviceId);
+                return commandConsumerHandler;
+            }
+        }
+        return commandConsumerHandlers.get(gatewayOrDeviceId);
+    }
+
+    /**
+     * Gets the device ids of the contained command handlers.
+     *
+     * @return The set of device ids.
+     */
+    public Set<String> getHandledDevices() {
+        return commandConsumerHandlers.keySet();
+    }
+
+    private void onRemoteClose(final Handler<String> remoteCloseHandler, final String event) {
+        remoteCloseHandler.handle(event);
+        commandConsumerHandlers.values().forEach(commandConsumerHandler -> {
+            commandConsumerHandler.getRemoteCloseHandler().handle(null);
+        });
+    }
+
+    private void onLocalClose(final Handler<String> localCloseHandler, final String event) {
+        localCloseHandler.handle(event);
+        commandConsumerHandlers.values().forEach(commandConsumerHandler -> {
+            commandConsumerHandler.getLocalCloseHandler().handle(null);
```

Comment:
same here?
Also: wouldn't it improve encapsulation if the `CommandConsumerHandler` would instead have methods *closeRemotely* and *closeLocally* instead of exposing its internal structure?